File: scrapegraph-js/examples/smartScraper_steps_example.js
import { smartScraper } from 'scrapegraph-js';
import 'dotenv/config';

/**
 * Example demonstrating how to use the SmartScraper with interactive steps.
 * This example shows how to navigate websites like a human user before extracting data.
 * 
 * Interactive steps allow you to:
 * - Click on elements
 * - Fill input fields
 * - Wait for page loads
 * - Navigate through multiple pages
 * - Perform complex user interactions
 */

// Configuration
const apiKey = process.env.SGAI_APIKEY;
const url = 'https://github.com/';
const prompt = 'Extract user profile information';

// Interactive steps for website navigation
const steps = [
  'click on search bar',
  'wait for 500ms',
  'fill email input box with mdehsan873@gmail.com',
  'wait a sec',
  'click on the first result of search',
  'wait for 2 seconds to load the result of search'
];

console.log('🚀 Starting SmartScraper with Interactive Steps...');
console.log(`🌐 Website URL: ${url}`);
console.log(`🎯 User Prompt: ${prompt}`);
console.log(`📋 Interactive Steps: ${steps.length} steps configured`);
console.log('\n' + '='.repeat(60));

// Display interactive steps
console.log('🎯 Interactive Steps to Execute:');
steps.forEach((step, index) => {
  console.log(`  ${index + 1}. ${step}`);
});
console.log('\n' + '='.repeat(60));

// Start timer
const startTime = Date.now();
console.log(`⏱️  Timer started at: ${new Date(startTime).toLocaleTimeString()}`);
console.log('🔄 Processing request with interactive steps...');

try {
  // Make request with interactive steps
  const response = await smartScraper(apiKey, url, prompt, null, null, steps);
  
  // Calculate execution time
  const endTime = Date.now();
  const executionTime = (endTime - startTime) / 1000;
  const executionMinutes = executionTime / 60;
  
  console.log(`⏱️  Timer stopped at: ${new Date(endTime).toLocaleTimeString()}`);
  console.log(`⚡ Total execution time: ${executionTime.toFixed(2)} seconds (${executionMinutes.toFixed(2)} minutes)`);
  console.log(`📊 Performance: ${executionTime.toFixed(1)}s (${executionMinutes.toFixed(1)}m) for ${steps.length} interactive steps`);
  
  // Display results
  console.log('✅ Request completed successfully!');
  console.log(`📊 Request ID: ${response.request_id || 'N/A'}`);
  console.log(`🔄 Status: ${response.status || 'N/A'}`);
  
  if (response.error) {
    console.log(`❌ Error: ${response.error}`);
  } else {
    console.log('\n📋 EXTRACTED DATA:');
    console.log('='.repeat(60));
    
    // Pretty print the result
    if (response.result) {
      console.log(JSON.stringify(response.result, null, 2));
      
      // Display extraction statistics
      console.log('\n📊 EXTRACTION STATISTICS:');
      console.log('-'.repeat(50));
      const resultStr = JSON.stringify(response.result);
      console.log(`📝 Data size: ${resultStr.length} characters`);
      console.log(`🔗 JSON keys: ${typeof response.result === 'object' ? Object.keys(response.result).length : 'N/A'}`);
      console.log(`⚡ Processing speed: ${Math.round(resultStr.length / executionTime)} chars/second`);
      console.log(`🎯 Steps efficiency: ${(executionTime / steps.length).toFixed(2)}s per step`);
    } else {
      console.log('No result data found');
    }
  }
  
} catch (error) {
  const endTime = Date.now();
  const executionTime = (endTime - startTime) / 1000;
  const executionMinutes = executionTime / 60;
  
  console.log(`⏱️  Timer stopped at: ${new Date(endTime).toLocaleTimeString()}`);
  console.log(`⚡ Execution time before error: ${executionTime.toFixed(2)} seconds (${executionMinutes.toFixed(2)} minutes)`);
  console.log(`💥 Error occurred: ${error.message}`);
  console.log('\n🛠️  Troubleshooting:');
  console.log('1. Make sure your .env file contains SGAI_APIKEY');
  console.log('2. Check your internet connection');
  console.log('3. Verify the target website is accessible');
  console.log('4. Ensure you have sufficient credits in your account');
}

// Additional examples of different step patterns
console.log('\n🎯 ADDITIONAL STEP PATTERN EXAMPLES:');
console.log('='.repeat(60));

const stepPatterns = [
  {
    name: 'Authentication Flow',
    description: 'Steps for logging into a website',
    steps: [
      'click on login button',
      'wait for 1 second',
      'fill username field with user@example.com',
      'wait for 200ms',
      'fill password field with password123',
      'wait for 300ms',
      'click submit button',
      'wait for 3 seconds'
    ]
  },
  {
    name: 'Form Interaction',
    description: 'Steps for filling and submitting a form',
    steps: [
      'scroll to contact form',
      'wait for 500ms',
      'fill name field with John Doe',
      'wait for 200ms',
      'fill email field with john@example.com',
      'wait for 200ms',
      'fill message field with Hello World',
      'wait for 300ms',
      'click submit button'
    ]
  },
  {
    name: 'Dynamic Content Loading',
    description: 'Steps for loading more content dynamically',
    steps: [
      'scroll to bottom of page',
      'wait for 1 second',
      'click load more button',
      'wait for 2 seconds',
      'scroll down again',
      'wait for 1 second',
      'click show details button'
    ]
  }
];

stepPatterns.forEach((pattern, index) => {
  console.log(`\n📋 Pattern ${index + 1}: ${pattern.name}`);
  console.log(`📝 Description: ${pattern.description}`);
  console.log(`🎯 Steps (${pattern.steps.length}):`);
  pattern.steps.forEach((step, stepIndex) => {
    const stepType = step.includes('click') ? 'Navigation' : 
                     step.includes('wait') ? 'Wait' : 
                     step.includes('fill') ? 'Input' : 
                     step.includes('scroll') ? 'Action' : 'Other';
    console.log(`   ${stepIndex + 1}. ${step} [${stepType}]`);
  });
  console.log('-'.repeat(40));
});

console.log('\n🎯 SMART SCRAPER INTERACTIVE STEPS EXAMPLE COMPLETED');
console.log('='.repeat(60));
console.log('This example demonstrates how to use interactive steps with SmartScraper.');
console.log('Interactive steps allow you to navigate websites like a human user.');
console.log('This enables more sophisticated data extraction from complex web applications.'); 

File: scrapegraph-py/examples/sync/steps/smartscraper_steps_example.py
"""
Example demonstrating how to use the Smart Scraper API with interactive steps.
This example shows how to:
1. Set up interactive steps for website navigation
2. Use the Smart Scraper API with custom steps
3. Handle the response and display results with timing
4. Demonstrate the power of interactive movements

Interactive steps allow you to:
- Click on elements
- Fill input fields
- Wait for page loads
- Navigate through multiple pages
- Perform complex user interactions

Requirements:
- Python 3.7+
- scrapegraph-py
- A .env file with your SGAI_API_KEY

Example .env file:
SGAI_API_KEY=your_api_key_here
"""

import json
import os
import time
from dotenv import load_dotenv
from scrapegraph_py import Client

# Load environment variables from .env file
load_dotenv()


def smart_scraper_with_steps():
    """
    Enhanced Smart Scraper function with interactive steps and comprehensive timing.
    This demonstrates how to use interactive movements to navigate websites.
    """
    
    # Get API key from environment
    api_key = os.getenv("SGAI_API_KEY")
    if not api_key:
        raise ValueError(
            "API key must be provided or set in .env file as SGAI_API_KEY. "
            "Create a .env file with: SGAI_API_KEY=your_api_key_here"
        )

    # Interactive steps for website navigation
    steps = [
        "click on search bar",
        "wait for 500ms",
        "fill email input box with mdehsan873@gmail.com",
        "wait a sec",
        "click on the first result of search",
        "wait for 2 seconds to load the result of search",
    ]

    # Target website configuration
    website_url = "https://github.com/"
    user_prompt = "Extract user profile information"

    print("🚀 Starting Smart Scraper with Interactive Steps...")
    print(f"🌐 Website URL: {website_url}")
    print(f"🎯 User Prompt: {user_prompt}")
    print(f"📋 Interactive Steps: {len(steps)} steps configured")
    print("\n" + "=" * 60)

    # Display interactive steps
    print("🎯 Interactive Steps to Execute:")
    for i, step in enumerate(steps, 1):
        print(f"  {i}. {step}")
    print("\n" + "=" * 60)

    # Start timer
    start_time = time.time()
    print(f"⏱️  Timer started at: {time.strftime('%H:%M:%S', time.localtime(start_time))}")
    print("🔄 Processing request with interactive steps...")

    try:
        # Initialize client
        with Client.from_env() as client:
            # Make request with interactive steps
            response = client.smartscraper(
                user_prompt=user_prompt,
                website_url=website_url,
                steps=steps
            )

            # Calculate execution time
            end_time = time.time()
            execution_time = end_time - start_time
            execution_minutes = execution_time / 60

            print(f"⏱️  Timer stopped at: {time.strftime('%H:%M:%S', time.localtime(end_time))}")
            print(f"⚡ Total execution time: {execution_time:.2f} seconds ({execution_minutes:.2f} minutes)")
            print(f"📊 Performance: {execution_time:.1f}s ({execution_minutes:.1f}m) for {len(steps)} interactive steps")

            # Display results
            print("✅ Request completed successfully!")
            print(f"📊 Request ID: {response.get('request_id', 'N/A')}")
            print(f"🔄 Status: {response.get('status', 'N/A')}")

            if response.get("error"):
                print(f"❌ Error: {response['error']}")
            else:
                print("\n📋 EXTRACTED DATA:")
                print("=" * 60)
                
                # Pretty print the result
                if "result" in response:
                    result_data = response["result"]
                    print(json.dumps(result_data, indent=2, ensure_ascii=False))
                    
                    # Display extraction statistics
                    print("\n📊 EXTRACTION STATISTICS:")
                    print("-" * 50)
                    result_str = json.dumps(result_data)
                    print(f"📝 Data size: {len(result_str)} characters")
                    print(f"🔗 JSON keys: {len(result_data) if isinstance(result_data, dict) else 'N/A'}")
                    print(f"⚡ Processing speed: {len(result_str)/execution_time:.0f} chars/second")
                    print(f"🎯 Steps efficiency: {execution_time/len(steps):.2f}s per step")
                else:
                    print("No result data found")

    except Exception as e:
        end_time = time.time()
        execution_time = end_time - start_time
        execution_minutes = execution_time / 60
        
        print(f"⏱️  Timer stopped at: {time.strftime('%H:%M:%S', time.localtime(end_time))}")
        print(f"⚡ Execution time before error: {execution_time:.2f} seconds ({execution_minutes:.2f} minutes)")
        print(f"💥 Error occurred: {str(e)}")


def smart_scraper_multiple_scenarios():
    """
    Demonstrate different interactive step scenarios.
    """
    print("\n🎯 MULTIPLE SCENARIOS DEMONSTRATION")
    print("=" * 60)
    
    scenarios = [
        {
            "name": "Basic Search",
            "url": "https://github.com/",
            "prompt": "Extract repository information",
            "steps": [
                "click on search bar",
                "wait for 300ms",
                "fill search with 'python'",
                "wait for 1 second",
                "click first result"
            ]
        },
        {
            "name": "Form Interaction",
            "url": "https://scrapegraphai.com/",
            "prompt": "Extract contact information",
            "steps": [
                "scroll down to contact form",
                "wait for 500ms",
                "click on email field",
                "wait for 200ms",
                "fill email with test@example.com"
            ]
        },
        {
            "name": "Navigation Flow",
            "url": "https://github.com/",
            "prompt": "Extract user profile",
            "steps": [
                "click on profile menu",
                "wait for 1 second",
                "click on settings",
                "wait for 2 seconds",
                "scroll to profile section"
            ]
        }
    ]
    
    for i, scenario in enumerate(scenarios, 1):
        print(f"\n📋 Scenario {i}: {scenario['name']}")
        print(f"🌐 URL: {scenario['url']}")
        print(f"🎯 Prompt: {scenario['prompt']}")
        print(f"📝 Steps: {len(scenario['steps'])}")
        for j, step in enumerate(scenario['steps'], 1):
            print(f"    {j}. {step}")
        print("-" * 40)


def main():
    """
    Main function to run the Smart Scraper steps example.
    """
    try:
        print("🎯 SMART SCRAPER INTERACTIVE STEPS EXAMPLE")
        print("=" * 60)
        print("This example demonstrates how to use interactive steps with Smart Scraper.")
        print("Interactive steps allow you to navigate websites like a human user.")
        print()
        
        smart_scraper_with_steps()
        smart_scraper_multiple_scenarios()
        
    except Exception as e:
        print(f"💥 Error occurred: {str(e)}")
        print("\n🛠️  Troubleshooting:")
        print("1. Make sure your .env file contains SGAI_API_KEY")
        print("2. Check your internet connection")
        print("3. Verify the target website is accessible")
        print("4. Ensure you have sufficient credits in your account")


if __name__ == "__main__":
    main() 